routing/http/types/record_announcement.go

```diff
@@ -0,0 +1,368 @@
+package types
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/ipfs/boxo/util"
+	"github.com/ipfs/go-cid"
+	"github.com/ipld/go-ipld-prime"
+	"github.com/ipld/go-ipld-prime/codec/dagjson"
+	"github.com/ipld/go-ipld-prime/datamodel"
+	"github.com/ipld/go-ipld-prime/node/basicnode"
+	"github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/multiformats/go-multibase"
+)
+
+const SchemaAnnouncement = "announcement"
+const announcementSignaturePrefix = "PUT /routing/v1 announcement:"
+
+var _ Record = &AnnouncementRecord{}
+
+type AnnouncementScope string
+
+const (
+	AnnouncementBlock     AnnouncementScope = "block"
+	AnnouncementEntity    AnnouncementScope = "entity"
+	AnnouncementRecursive AnnouncementScope = "recursive"
+)
+
+// AnnouncementRequest represents a request to announce
+type AnnouncementRequest struct {
+	CID      cid.Cid
+	Scope    AnnouncementScope
+	TTL      time.Duration
+	Metadata []byte
+}
+
+// AnnouncementPayload is the [AnnouncementRecord] payload.
+type AnnouncementPayload struct {
+	CID       cid.Cid
+	Scope     AnnouncementScope
+	Timestamp time.Time
+	TTL       time.Duration
+	ID        *peer.ID
+	Addrs     []Multiaddr
+	Protocols []string
+	Metadata  string
+}
+
+// MarshalJSON marshals the [AnnouncementPayload] into a canonical DAG-JSON
+// representation of the payload.
+func (ap AnnouncementPayload) MarshalJSON() ([]byte, error) {
+	m := make(map[string]ipld.Node)
+	var err error
+
+	// TODO: or goes empty? Affects signature.
+	if ap.CID.Defined() {
+		m["CID"] = basicnode.NewString(ap.CID.String())
+	}
+
+	// TODO: is scope not set when empty? Affects signature.
```

Comment:
We should skip fields when unset or empty. `""` is a golang idiom, but not that obvious in other languages.